### ISSUE TITLE
[Help needed] Weird prompt on OS X

### ISSUE THREAD

**Issue by sbstp** [kubie-org/kubie#13]
I've received reports of the bash prompt being weird inside of a kubie shell in OS X, currently investigating the issue.

Instead of using the default user `PS1` it uses a weird `PS1` that shows bash's version.

It looks like: `[context|namespace] bash-3.2$ `

I'd like to collect more info about this because I don't own a Mac. If anyone experiences this problem, please let me know.

**Comment by tbondarchuk** [comment]
Just reproduced it if no .bashrc is present or PS1 is not configured in .bashrc.

MacOS Catalina switched to using zsh as default shell so I guess on new installs, not upgrades like mine, .bashrc is just not present there. Easy fix will be simply to create .bashrc with PS1 configured

Or update kubie to support zsh/current shell :) Personally I can work with kubie using bash at the moment, as in theory I would need to run only kubectl in spawned bash shell, but for some it might cause inconveniences as all aliases, etc from .zshrc will be gone, as well as custom PS1.

Very nice tool, many thanks! I've only recently started using kubectx/kubens but I like kubie more :) Waiting for brew package and zsh support, but started using it already. 

Please let me know if I can help with debugging on MacOS

P.S. bit of weird and off top request, are there any way to disable kubectx/kubens aliases in spawned shell? I haven't developed muscle memory yet and it's kinda bugging me that I can't quickly autocomplete `kubectl` command as `kubectx` stands in a way.

**Comment by sbstp** [comment]
Ah yes I see. The long game is definitely supporting `zsh`. I was hoping for a temporary fix, but if bash isn't even installed, I'll put more focus on zsh!

**Comment by sbstp** [comment]
Zsh support has been released, which should fix the weird bash related issues, since zsh will be used instead.

**Comment by sryabkov** [comment]
Plenty of people use bash on a Mac even after the switch to `zsh`. In fact, I didn't even realize the switch happened when I upgraded to Catalina as my shell setting were preserved during the upgrade. [I use a version of bash that is installed with `homebrew`](https://itnext.io/upgrading-bash-on-macos-7138bd1066ba).

When you start a terminal on a Mac, [`~/.bash_profile` gets loaded, not `~/.bashrc`](https://apple.stackexchange.com/questions/51036/what-is-the-difference-between-bash-profile-and-bashrc). For that reason, all my shell customizations, aliases, etc are in `.bash_profile`. But if you start another shell from an existing shell (e.g. run `bash` from command line), `.bashrc` gets loaded. This is what seems to happen when `kubie` is executed.

I also get the "weird prompt" when I run `kubie ctx` (`[context|namespace] bash-5.0$`), and none of my settings are loaded from `.bash_profile`.

**Comment by sbstp** [comment]
Yeah according to [this](https://apple.stackexchange.com/a/51038) mac starts a login shell, which is why `bash_profile` is used instead of `bashrc`. I also noticed this with my shell detection feature.

Definitely something we can fix.

**Comment by uhlhosting** [comment]
Got same issue on Cataline, it moves to bash shell when running it.

**Comment by sbstp** [comment]
@sryabkov sorry I didn't read your comment properly when I first saw it. You linked to the same stackoverflow issue that I found. You nailed it, it's an issue with login shells. 

I opened a merge request with a potential fix, I will test it tomorrow with my OS X VM.

**Comment by sbstp** [comment]
Released version 0.8.3 with this fix!

**Comment by sryabkov** [comment]
Thank you! The good news is that my `~/bash_profile` is definitely loaded now, the bad news is that the "weird prompt" (` ([context|namespace] bash-5.0$)`) is still there. 

I dug a little deeper into the issue, and profile loading seems to be a lot more complex than what kubie is trying to do in #28 (and which I missed myself during the code review yesterday). As the Invocation section of [`man bash`](https://linux.die.net/man/1/bash)  reads:

> When **bash** is invoked as an interactive login shell, or as a non-interactive shell with the **--login** option, it first reads and executes commands from the file _/etc/profile_, if that file exists. After reading that file, it looks for _~/.bash_profile, ~/.bash_login, and ~/.profile_, in that order, and reads and executes commands from the first one that exists and is readable. The **--noprofile** option may be used when the shell is started to inhibit this behavior.

Note: emphasis above is by `man bash`

Below is what seems to be happening on my Mac.

1. `/etc/profile` is loaded first:

```
# System-wide .profile for sh(1)

if [ -x /usr/libexec/path_helper ]; then
	eval `/usr/libexec/path_helper -s`
fi

if [ "${BASH-no}" != "no" ]; then
	[ -r /etc/bashrc ] && . /etc/bashrc
fi
```

2. `/etc/bashrc` is loaded next (and that is the file that sets `PS1`):

```
# System-wide .bashrc file for interactive bash(1) shells.
if [ -z "$PS1" ]; then
   return
fi

PS1='\h:\W \u\$ '
# Make bash check its window size after a process completes
shopt -s checkwinsize

[ -r "/etc/bashrc_$TERM_PROGRAM" ] && . "/etc/bashrc_$TERM_PROGRAM"
```
3. My `~/.bash_profile` seems to get loaded only after the above two files are.

> Note: I never customized the above two files (`/etc/profile` and `/etc/bashrc`) on my Mac, but my machine is over 3 years old. It originally came with whatever version of OSX was current in late 2016, and went through three OS upgrades. So, on a new machine, these files might potentially be different. Plus a fresh Mac would be configured to use `zsh` by default.

[Some people tried to organize and present all this logic of shell profile loading in a more digestable form than what `man bash` provides](https://shreevatsa.wordpress.com/2008/03/30/zshbash-startup-files-loading-order-bashrc-zshrc-etc/). 

In summary, if kubie wants to be perfect, there seems to be more work to do. If not, the fix in #28 is good enough. 

(edited for clarity on 2020-04-12)

**Comment by sbstp** [comment]
Thank you for the investigation! I want to do this properly so I'll look at doing this better!

**Comment by uhlhosting** [comment]
Some issues around caching connections needs to be debugged. kubie somehow can keep a conection active even when a new exporrt is made with new config. would be great to trigger a change in kubie once any changes in the env vars.

**Comment by sbstp** [comment]
This is a feature of kubie. If you play with `KUBECONFIG` for some reason, the change to it does not affect the current shell. If you look at the generated bashrc or zshrc, you will see that we install a hook that runs before commands and sets `KUBIECONFIG=$KUBIE_KUBECONFIG`.

**Comment by sbstp** [comment]
Released version 0.8.4 which should hopefully improve bash's situation on Mac!

**Comment by sryabkov** [comment]
Just tried it, looks great. Thank you @sbstp!
